src/job_cache/message_sender.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2023 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#define _XOPEN_SOURCE 700
+#define _POSIX_C_SOURCE 200809L
+
+#include <unistd.h>
+
+#include <algorithm>
+#include <string>
+#include <vector>
+
+namespace job_cache {
+
+enum class MessageSenderState {
+  Continue,
+  StopSuccess,
+  StopFail,
+  Timeout,
+};
+
+// Message sender acts sort of like a Rust future, it holds the state needed
+// to keep performing a write even if it would block. This can be thought of as
+// dual to MessageParser. MessageSender will continue writing until it receives
+// EAGAIN/EWOULDBLOCK, or an error occurs. Thus it is a *must* that non-blocking
+// io be used. It is recomended that you use this with edge triggered polling so
+// that you only reattempt the write when you have to. Otherwise you will receive
+// writes that you did not want to receive.
+class MessageSender {
+ public:
+  std::string data;
+  std::string::iterator start;
+  time_t deadline;
+  int fd;
+  MessageSenderState state = MessageSenderState::Continue;
+
+ public:
+  MessageSender() = delete;
+  MessageSender(const MessageSender&) = delete;
+  MessageSender(MessageSender&& sender) {
```

Comment:
do we actually use/need the move constructor here?

Reply:
Yes, we need the move constructor in several places. 1) To store an object in a resizable container (e.g. a hash table) you have to be moveable. 2) We construct a message sender and then move construct it via `emplace` when we add a message sender to message_senders.